design/ceph/update-osds-in-parallel.md

```diff
@@ -0,0 +1,173 @@
+support large clusters, Rook should be able to update (and upgrade) multiple OSDs in parallel.
+
+
+## High-level requirements
+
+### Different update strategies
+Users will likely wish to specify different upgrade strategies. For many clusters, updating all OSDs
+on a node may be sufficient. For portable OSDs on PVCs, an OSD may not be associated directly with a
+node.
+
+Users may wish to update all OSDs in a particular failure domain at once, updating all OSDs in a
+domain before moving to the next domain. Some examples of failure domains are:
+ - by OSD (equivalent to the existing strategy)
+ - by node/host
+ - by rack
+ - by data center room
+ - by availability zone
+ - and many others
+
+### Ability to set a maximumum number of OSDs to be updated simultaneously
+In the worst (but unlikely) case, all OSDs which are updated for a given parallel update operation
+might fail to come back online after they are updated. Users may wish to limit the number of OSDs
+updated in parallel in order to avoid too many OSDs failing in this way.
+
+### Cluster growth takes precedence over updates
+Adding new OSDs to a cluster should occur as quickly as possible. This allows users to make use of
+newly added storage as quickly as possible, which they may need for critical applications using the
+underlying Ceph storage. In some degraded cases, adding new storage may be necessary in order to
+allow currently-running Ceph OSDs to be updated without experiencing storage cluster downtime.
+
+This does not necessarily mean that adding new OSDs needs to happen before updates.
+
+This prioritization might delay updates significantly since adding OSDs not only adds capacity to
+the Ceph cluster but also necessitates data rebalancing. Rebalancing generates data movement which
+needs to settle for updates to be able to proceed.
+
+### OSD updates should not starve other resources of updates
+For Ceph cluster with huge numbers of OSDs, Rook's process to update OSDs should not starve other
+resources out of the opportunity to get configuration updates.
+
+
+## Technical implementation details
+
+### Workflow
+Use the `ceph osd ok-to-stop <id>...` command to determine if it is safe to update multiple OSDs at
+once.
+
+1. Build an "update queue" of OSDs which need updated.
+2. Query `ok-to-stop` for a single OSD from the update queue. This represents the CRUSH hierarchy
+   bucket at the "osd" (or "device") level.
+3. If the previous operation reports that it safe to update, batch query `ok-to-stop` for all OSDs
+   that fall under the CRUSH bucket one level up from the current level.
+4. Repeat step 3 moving up the CRUSH hierarchy until one of the following two conditions:
+    1. The number of OSDs in the batch query is greater or equal to the maximum number of OSDs the
+       user allows to be updated in parallel, OR
+    2. It is no longer `ok-to-stop` all OSDs in the CRUSH bucket.
```

Comment:
I wonder if it makes sense here to put this logic in ceph.  e.g., ``ceph osd ok-to-stop $id --maximize`` could return a json list of which OSDs are safe to stop (including $id), or fail as before if $id is not ok to stop.  That way we can do the same thing to speed up cephadm upgrades.  This will be somewhat annoying to code up so doing it once would be nice.  WDYT?

Reply:
I would tend to agree. Do you think the Ceph mgr is a good place for that code, or in Ceph's C++ internals?

Reply:
Wherever ok-to-stop is implemented... in the mgr IIRC!  I can start on this

Reply:
👏 My hero!